Add Docker-based metrics monitoring stack with Prometheus and Grafana

kkfernandezTT opened the pull request:
### Ticket
  Closes #30768

  ### Problem description
  We needed a rapid way to bring up Prometheus + Grafana locally to test TT-Metal telemetry metrics. The solution needed to:
  - Run on macOS (and ideally Linux) via Docker
  - Be configured to scrape from designated remote host(s)
  - Provide an easy wrapper script to specify which hosts to scrape from (e.g., `http://sjc-wh-05:54168/api/metrics`)

  ### What's changed
  Added a complete Docker Compose monitoring stack for TT-Metal telemetry with Prometheus and Grafana:

  **New files:**
  - `docker-compose.yml` - Defines Prometheus (port 9090) and Grafana (port 3000) services with persistent storage
  - `start-monitoring.sh` - Wrapper script that:
    - Accepts one or more host:port targets as arguments
    - Automatically resolves hostnames to IPv4 addresses (with fallback if resolution fails)
    - Generates `prometheus.yml` dynamically with the specified targets
    - Starts the monitoring stack via Docker Compose
  - `grafana-datasources.yml` - Auto-provisions Prometheus as a Grafana datasource
  - `README.md` - Complete documentation with setup, usage, and troubleshooting

  **Key features:**
  - **Simple usage**: `./start-monitoring.sh sjc-wh-05:53494 sjc-wh-06:53494` to monitor multiple hosts
  - **Hostname resolution**: Script attempts to resolve hostnames to IPs automatically (useful when Docker can't resolve internal hostnames)
  - **Enhanced labels**: Metrics are automatically labeled with `instance` (host:port), `host`, and `port` for flexible querying in Grafana
  - **Cross-platform**: Works on macOS and Linux using `host.docker.internal` / `host-gateway`
  - **Persistent data**: Uses Docker volumes to preserve Prometheus time-series data and Grafana dashboards across restarts

  **Impact:**
  Developers can now quickly spin up a local monitoring environment to visualize TT-Metal telemetry metrics from any remote host(s) in under 30, without manual Prometheus/Grafana configuration.

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

kkfernandezTT committed:
Extend WebServer functionality by exposing /api/metrics EP w/ metrics in Prometheus format

kkfernandezTT committed:
Add Docker Compose setup for Prometheus and Grafana monitoring

kkfernandezTT committed:
Satisfy GitHub Copilot lints

kkfernandezTT committed:
Simplify monitoring script to use unresolved hostnames and deduplicate targets

btrzynadlowski-tt commented on tt_telemetry/server/prom_formatter.cpp:
```diff
@@ -0,0 +1,236 @@
+    localtime_r(&now_c, &tm_buf);
+
+    output << "# Tenstorrent Metal Telemetry Metrics\n";
+    output << "# Host: " << hostname << "\n";
```
Do we need this? How does Prometheus use this information? We will be usually be scraping from the individual collectors (tt_telemetry run in its default collector mode, monitoring only the host it is running on). *But*, it is *possible* that tt_telemetry is run in aggregator mode, and aggregates metrics itself from a number of collectors. I guess this host would represent the host that the metrics are scraped from (but not necessarily the host that the metrics came from).

Elsewhere, you pass this down and assume you will find it in the metric path name. But what if it is running in aggregator mode? Metrics from multiple hosts will be there.

btrzynadlowski-tt reviewed (changes requested):
Looks good but the hostname issue definitely needs to be fixed. Cannot assume that a metric came from the same host.

kkfernandezTT committed:
Address reviewer comments

- Replace C header limits.h with climits
- Extract hostname dynamically from metric paths for aggregator mode compatibility
- Remove misleading hostname comment from Prometheus output
- Fix macOS bash 3 compatibility in start-monitoring.sh

kkfernandezTT committed:
Add hostname resolution for macOS Docker compatibility and update to host networking

btrzynadlowski-tt reviewed (approved):
Great job!

kkfernandezTT committed:
Move monitoring stack to dedicated subdirectory

tt-rkim reviewed (commented):
I LOVE IT!!!!!
